baby-bear/src/mds.rs

```diff
@@ -0,0 +1,553 @@
+// --(4)                      >= (2**N)*input - (2**N)*input_low - (2**N)*input_high   (Assuming P < 2**N)
+//
+// Dividing by 2**N we get
+// Q*P + 2^{10}*P >= input - input_low - input_high
+// which rearranges to
+// x' = input - Q*P <= 2^{10}*P + input_low + input_high
+//
+// Picking N = 40 we see that 2^{10}*P, input_low, input_high are all bounded by 2**40
+// Hence x' < 2**42 < 2**50 as desired.
+//
+//
+//
+///////////////////////////////////////////////////////////////////////////
+// CASE 2: input < 0
+//
+// This case will be similar but all our inequalities will change slightly as negatives complicate things.
+// First observe that:
+// (input >> N) * I   >= (input >> N) * 2**(2N) / P
+//                    >= (1 + (input / 2**N)) * 2**(2N) / P
+//                    >= (2**N + input) * 2**N / P
+//
+// Thus:
+// Q = ((input >> N) * I) >> N >= ((2**N + input) * 2**N / P) >> N
+//                             >= ((2**N + input) / P) - 1
+//
+// And so sub = Q*P >= 2**N - P + input.
+// Hence input - sub < 2**N - P.
+//
+// Thus if input - sub > 0 then |input - sub| < 2**50.
+// Thus we are left with bounding -(input - sub) = (sub - input).
+// Again we will proceed by improving our bound on Q.
+//
+// Q = (((input_high * I) >> N) & mask)
+// --(2)   => Q <= (input_high * I) >> N) <= (I*x_high)/(2**N)
+// --(1)   => Q <= (I*x_high)/(2**N)
+//         => (2**N)*Q <= I*x_high
+//
+// Hence we find that:
+// (2**N)*Q*P <= input_high*I*P
+// --(3a)     <= input_high*2**{2N} - P*input_high
+// --(4)      <= (2**N)*input - (2**N)*input_low - (2**N)*input_high   (Assuming P < 2**N)
+//
+// Dividing by 2**N we get
+// Q*P <= input - input_low - input_high
+// which rearranges to
+// -x' = -input + Q*P <= -input_high - input_low < 2**50
+//
+// This completes the proof.
+
+/// Instantiate convolution for "large" RHS vectors over BabyBear.
+///
+/// Here "large" means the elements can be as big as the field
+/// characteristic, and the size N of the RHS is <= 64.
+struct LargeConvolveBabyBear;
+impl Convolve<BabyBear, i64, i64, i64> for LargeConvolveBabyBear {
+    /// Return the lift of a BabyBear element, satisfying 0 <=
+    /// input.value < P < 2^31. Note that BabyBear elements are
+    /// represented in Monty form.
+    #[inline(always)]
+    fn read(input: BabyBear) -> i64 {
+        input.value as i64
+    }
+
+    /// For a convolution of size N, |x|, |y| < N * 2^31, so the
+    /// product could be as much as N^2 * 2^62. This will overflow an
+    /// i64, so we first widen to i128. Note that N^2 * 2^62 < 2^80
+    /// for N <= 64, as required by `barret_red_babybear()`.
+    #[inline(always)]
+    fn parity_dot<const N: usize>(u: [i64; N], v: [i64; N]) -> i64 {
+        let mut dp = 0i128;
+        for i in 0..N {
+            dp += u[i] as i128 * v[i] as i128;
+        }
+        barret_red_babybear(dp)
+    }
+
+    /// The assumptions above mean z < N^3 * 2^62, which is at most
```

Comment:
Need to fix the comments 212 - 215 and 225-233.

Current comments are before we implemented the partial reduction method.

Reply:
I think the logic for the bound for the input to reduce should be:

After the barret reduction method, the output z of parity dot satisfies |z| < 2^50 (See Thm 1 on line 95).

In the recombining steps, conv_n maps (wo, w1) -> ((wo + w1)/2, (wo + w1)/2) which has no effect on the maximal size. (Indeed, it makes sizes almost strictly smaller).

On the other hand, signed_conv_n (ignoring the re-index) recombines as: (w0, w1, w2) -> (w0 + w1, w2 - w0 - w1). Hence if the input is <= K, the output is <= 3K.

Thus the values appearing at the end are bounded by 3^n 2^50 where n is the maximal number of signed_conv recombination steps. When N = 64, we need to recombine for singed_conv_32, singed_conv_16, singed_conv_8 so the overall bound will be 3^3 2^50 < 32 * 2^50 < 2^55.